### DISCUSSION TITLE
Proper versioning for proper deployment (frontend/backend version syncing)

### DISCUSSION THREAD

**Discussion by must** [loreanvictor/tyfon#18] [Ideas]
In the context of a production environment, and when deploying different frontend and backend versions, we should find a way to handle these transitions. This can occur in multiple forms:
- The user has an older client version and hence might have to use an older corresponding backend version (but this does not have to be the case since a newer version can contain bug fixes and no breaking API changes).
- When deploying a new frontend version, it might not be guaranteed that the new backend version is deployed at the same time (can also be different for different users at any instant as the transition is not instantaneous). There will always be a lag time for when the infrastructure routes requests to the new version as opposed to the older one. This becomes more important the more traffic the backend gets.
- Version rollbacks on the backend (e.g. to recover from a faulty deployment) and by extension in some cases for the frontend.
- Some advanced use cases might imply slowly rolling out updates and checking for potential errors and breaking changes on production. A/B testing of sorts. This will be important for more mature production environments depending on the use case.

**Comment by loreanvictor** [comment]
so I turned this into a discussion as it is not about an actual issue on TyFON that can be resolved (or a feature that can be added). of course, issues might come out of this discussion.

---

<br>

> - The user has an older client version and hence might have to use an older corresponding backend version (but this does not have to be the case since a newer version can contain bug fixes and no breaking API changes).

Right now, TyFON offers type checks which will automatically catches breaking API changes on build time. These are the scenarios we could do something more in a meaningful way:

<br>

### There is a version of the API compatible with requirements of the client somewhere.

👉 TyFON would need to know about other versions, the URLs for accessing them, etc. The easiest way would be for each instance of the API to point to a _version registry_. This registry can be a simple URL providing information about which versions of the API are available and at which URLs (and potentially until when).

The TyFON client then would need to keep two URLs per SDK: a reference URL, and a resolved URL. The reference URL is where it will always check for updates and reference to version registry (and is the URL the user provides), while resolved URL refers to the actual API url where SDK metadata should be retrieved from and where requests should be sent to.

<br>

### There is a breaking change without syntactic errors

This means some function has changed functionality in a breaking way, but without a signature change. This can happen, but I assumed it is pretty rare.

👉 TyFON would need to check for version compatibility generally using [semver](https://semver.org), as checking semantic correctness is prohibitively costly. This would resolve the current issue as well.

Perhaps generally TyFON needs to check versions and at least warn about version mismatch. In case a _version registry_ is specified on the API, it should try to find a more suitable version based on that info, otherwise it should try to install regardless, hoping that version mismatch won't cause issues (which might be the case if semantic versioning is not followed properly for the API, which can happen a lot).

<br>

---

<br>

> - When deploying a new frontend version, it might not be guaranteed that the new backend version is deployed at the same time 

Current response from TyFON seems like the best option to this case. The client requires a version of the server that is not deployed, the only correct response is to fail the client build.

Of course, there is the corner case where the client needs to fail since there is a breaking change without a breaking signature change. As discussed above, this is rare and would be covered by the proposed semver checks.

<br>

---

<br>

> - Version rollbacks on the backend (e.g. to recover from a faulty deployment) and by extension in some cases for the frontend.

This is completely up to production environments and feels out of scope for a discussion about TyFON. That said, I guess a proper production environment would enable webhooks and CI checks for version rollout / rollback in general: For example you can have a service grid where every rollout / rollback is announced to all active containers, where a side car checks whether this rollout / rollback would negatively affect the running service. In case it is and it can rebuild with a different available version, then it would do so, otherwise it would need to send a signal to fail the rollout / rollback process (as some part of the cluster would stop operating as a result of that action). Note that this is just a rough idea and perhaps this idea on its own would have some negative impact on efficiency of CI / CD pipelines as every running container / service would need to green-light every rollout / rollback.

<br>

---

<br>

> - Some advanced use cases might imply slowly rolling out updates and checking for potential errors and breaking changes on production. A/B testing of sorts. This will be important for more mature production environments depending on the use case.

Good point, but I think simply adding support for [dist-tags](https://docs.npmjs.com/adding-dist-tags-to-packages) alongside semver checks will resolve this issue (exactly the same way it works for NPM packages).

Note that these changes (semver and dist-tags based checks) are really reliant on two factors that are completely in developer's hands and cannot be guaranteed by TyFON itself:

- Semver **MUST** be followed strictly: breaking changes **MUST** have major version increases, non-breaking changes **MUST** have non-major version increases.
- Versions should be immutable: this is automatically true for NPM packages as package registries enforce that, i.e. once a package is published its content can't be modified unless the version changes. Similar requirement would apply to TyFON servers, but that is purely up to developers and the production cluster.

**Comment by loreanvictor** [comment]
I have created [a separate document](https://github.com/loreanvictor/immutable-cloud) for a general structure for solving API versioning. It might be a good idea to finalize that document first, then bring over any applicable tasks as issues to TyFON.
